plugins/zenoh-plugin-storage-manager/src/replica/storage.rs

```diff
@@ -186,35 +261,235 @@ impl StorageService {
+                let sample_to_store = match self
+                    .ovderriding_wild_update(&k, sample.get_timestamp().unwrap())
+                    .await
+                {
+                    Some(overriding_update) => {
+                        let mut sample_to_store =
+                            Sample::new(KeyExpr::from(k.clone()), overriding_update.data.value)
+                                .with_timestamp(overriding_update.data.timestamp);
+                        sample_to_store.kind = overriding_update.kind;
+                        sample_to_store
+                    }
+                    None => {
+                        let mut sample_to_store =
+                            Sample::new(KeyExpr::from(k.clone()), sample.value.clone())
+                                .with_timestamp(sample.timestamp.unwrap());
+                        sample_to_store.kind = sample.kind;
+                        sample_to_store
+                    }
+                };
+
+                let stripped_key = match self.strip_prefix(&sample_to_store.key_expr) {
+                    Ok(stripped) => stripped,
+                    Err(e) => {
+                        error!("{}", e);
+                        return;
+                    }
+                };
+                let mut storage = self.storage.lock().await;
+                let result = if sample.kind == SampleKind::Put {
+                    storage
+                        .put(
+                            stripped_key,
+                            sample_to_store.value.clone(),
+                            sample_to_store.timestamp.unwrap(),
+                        )
+                        .await
+                } else if sample.kind == SampleKind::Delete {
+                    // register a tombstone
+                    self.mark_tombstone(&k, sample_to_store.timestamp.unwrap())
+                        .await;
+                    storage
+                        .delete(stripped_key, sample_to_store.timestamp.unwrap())
+                        .await
+                } else {
+                    Err("sample kind not implemented".into())
+                };
+                if self.replication.is_some()
+                    && result.is_ok()
+                    && !matches!(result.unwrap(), StorageInsertionResult::Outdated)
+                {
+                    let sending = self
+                        .replication
+                        .as_ref()
+                        .unwrap()
+                        .log_propagation
+                        .send((k.clone(), *sample_to_store.get_timestamp().unwrap()));
+                    match sending {
+                        Ok(_) => (),
+                        Err(e) => {
+                            error!("Error in sending the sample to the log: {}", e);
+                        }
+                    }
+                }
+                drop(storage);
```

Comment:
Couldn't this drop be called earlier, just before `if self.replication.is_some()` ?
This would minimise the lock time.